Spigot-Server-Patches/0333-Add-entity-count-cache.patch

```diff
@@ -0,0 +1,148 @@
+         {
+             guard();
++            if(o instanceof Entity) updateEntityCount((Entity)o, false); // Paper
+             return super.remove( o );
+         }
+ 
+@@ -82,6 +114,7 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+     public final Map<String, EntityHuman> playersByName = Maps.newHashMap(); // Paper - World EntityHuman Lookup Optimizations
+     public final List<Entity> k = Lists.newArrayList();
+     protected final IntHashMap<Entity> entitiesById = new IntHashMap();
++    private Map<Class<?>, Integer> countCache = new HashMap(); // Paper
+     private final long G = 16777215L;
+     private int H; public int getSkylightSubtracted() { return this.H; } public void setSkylightSubtracted(int value) { this.H = value;} // Paper - OBFHELPER
+     protected int m = (new Random()).nextInt();
+@@ -2439,19 +2472,16 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+ 
+     }
+ 
+-    public int a(Class<?> oclass) {
++    // Paper start - entity count cache
++    private int countEntityType(Class<?> oclass) {
+         int i = 0;
+         Iterator iterator = this.entityList.iterator();
+ 
+         while (iterator.hasNext()) {
+             Entity entity = (Entity) iterator.next();
+-            if (entity.shouldBeRemoved) continue; // Paper
+             // CraftBukkit start - Split out persistent check, don't apply it to special persistent mobs
+-            if (entity instanceof EntityInsentient) {
+-                EntityInsentient entityinsentient = (EntityInsentient) entity;
+-                if (entityinsentient.isTypeNotPersistent() && entityinsentient.isPersistent()) {
+-                    continue;
+-                }
++            if (shouldIgnoreForCount(entity)) {
++                continue;
+             }
+ 
+             if (oclass.isAssignableFrom(entity.getClass())) {
+@@ -2464,6 +2494,52 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         return i;
+     }
+ 
++    public int getEntityCount(Class<?> oclass) { return a(oclass); } // Paper - OBFHELPER
++    public int a(Class<?> oclass) {
++        if(countCache.containsKey(oclass)) {
++            return countCache.get(oclass);
++        } else {
++            int count = countEntityType(oclass);
++
++            countCache.put(oclass, count);
++
++            return count;
++        }
++    }
++
++    public boolean shouldIgnoreForCount(Entity entity) {
++        if (entity instanceof EntityInsentient) {
++            EntityInsentient entityinsentient = (EntityInsentient) entity;
++            return entityinsentient.isTypeNotPersistent() && entityinsentient.isPersistent();
++        }
++
++        return false;
++    }
++
++    protected void updateEntityCount(Entity entity, boolean incr) {
++        if(entity == null) return;
++        if(shouldIgnoreForCount(entity)) return;
++
++        countCache.replaceAll((clazz, count) -> {
```

Comment:
why replaceAll? we need to instead walk the super chain and insert by supers instead.

we can updateEntityCount(Class, boolean), then loop over the super chain until Entity is hit, then modify each.